enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,476 @@
+  - *Output record*: concrete record encoded by an `ouput` for a consumer. Encodings include JSON and Rsyslog.
+- *Schema*: Defines key names and value types in a structured entry.
+   The forwarder associates *schema names* with logs, it does not store or process user schema.
+- *Data Model*: The [schema][data_model] for the forwarder's own *record*.
+
+## Motivation
+
+### Goals
+
+* Direct access to JSON log entries as JSON objects in log records for `fluentdForward`, `Elasticsearch` or any other JSON-parsing consumer.
+* User can associate schema names with logs by category, namespace, application label, or any future forwarder `input` selection criteria.
+* Elasticsearch indexing of JSON logs with multiple schema without index explosions.
+* Upgrade path from today's implementation.
+
+### Non-Goals
+
+* General-purpose JSON queries and transformations.
+* Recording or validating user-defined schema.
+  The forwarder only identifies schema by name, all schema knowledge belongs to the user.
+
+## Proposal
+
+#### Data model
+
+Two new top-level fields added to the record [data model][data_model]:
+
+* `structured` (type object): Original structured JSON log entry.
+* `schema` (type string): Schema name assigned by user.
+
+Relationship between `structured` and `message` fields:
+
+* EXACTLY ONE of `message` or `structured` MUST be present.
+* `schema` is OPTIONAL, defined and interpreted by the user.
+* The presence or absence of the `structured` field can be used to detect structured/unstructured records.
+* If a filter (see below) removes the `structured` field, an equivalent `message` field is added by.
+
+With an *Elasticsearch* store, `schema` is copied to the existing `viaq_index_name` field to set the index.
+`viaq_index_name` will continue to be used for compatibility and upgrade, but will eventually be removed.
+
+#### ClusterLogForwarder configuration
+
+The elements of the CLF API are:
+
+- *Input*: Selects logs, adds meta-data, creates _internal_ records.
+- *Pipeline*: Transforms, filters and routes internal records.
+- *Output*: Connects to consumers, encodes and sends _output_ records.
+
+New fields:
+
+`input.stuctured`: (string, default "json") Attempt to parse structured entries.
+The only legal value is "json", others may be added in future.
+
+An input record contains:
+
+* `structured` field if `structured=json` *and* parse succeeds.
+* `message` field if `input.structured` is absent *or* JSON parse fails.
+
+`input.schema`: (string, optional) User-defined schema name.
+Adds a `schema` field to the internal record with a user-defined value.
+Not used by the forwarder, included in output record for the consumer.
+
+`pipeline.filter`: (object, optional) Remove unwanted records.
+
+* `reject`: (array of string) drop records that have any of the listed fields.
+* `require`: (array of string) drop records that _do not_ have all the listed fields.
+
+`pipeline.modify`: (object, optional) Modify record fields.
```

Comment:
so this seems like the proposal is that our forwarder will be doing the manipulation of records rather than leveraging (for example) Elasticsearch pipeline processing. Is this accurate?

Or is the idea that the pipeline would be configured to look for `message` or `structured` to determine which index it is valid for?

Reply:
The forwarder is doing some manipulation, elasticsearch will do more. 3 cases:

1. Non-elasticsearch target, provide some basic capabilities on forwarder.
2. Compatibility mode: if running new forwarder with old ES, the forwarder can set the `viaq_index_name==scheme` and we get pretty much the right behaviour of indexing by scheme (I think, please confirm) but still some risk of index explosion if user messages are badly behaved.
3. New ES pipelines: ES pipeline knows the data model and handle `structured`, `message` and `schema` properly. It can also take self-defense measures like limiting the depth of indexing on `structured`, not indexing `kubernetes.labels` etc.